src/tildePrefix.ml

```diff
@@ -37,46 +37,123 @@
 *)
 open CST
 
-(** Remove the ['~'] character at the beginning of the given string or fail with
-    [Invalid_arg] if the string is empty or does not start with a tilde. *)
-let strip_tilde s =
-  if s = "" || s.[0] != '~' then
+(** Tests whether the given string starts with a tilde character. *)
+let starts_with_tilde string =
+  string != "" && string.[0] = '~'
+
+(** Removes the ['~'] character at the beginning of the given string or fail
+    with [Invalid_arg] if the string does not start with a tilde. *)
+let strip_tilde string =
+  if not (starts_with_tilde string) then
     invalid_arg "strip_tilde";
-  String.(sub s 1 (length s - 1))
+  String.(sub string 1 (length string - 1))
 
-(** Takes a string assumed to come from a [WordLiteral] and to start with ['~']
-    and splits it between a [WordTildePrefix] containing everything between
-    ['~'] and the first ['/'] and a [WordLiteral] containing everything else. *)
-let find_login s =
-  match String.split_on_char '/' s with
-  | login :: rem ->
-    [
-      WordTildePrefix (strip_tilde login);
-      WordLiteral (String.concat "/" rem)
-    ]
-  | _ ->
+(** Extracts the tilde-prefix at the beginning of the given literal string or
+    fail with [Invalid_arg] if it does not start with a tilde. *)
+let extract_tilde_prefix_from_literal (literal : string) : word_cst =
+  if not (starts_with_tilde literal) then
+    invalid_arg "extract_tilde_prefix_from_literal";
+  match String.split_on_char '/' literal with
+  | [] ->
     (* [String.split_on_char] yields a list of at least one element. *)
     assert false
+  | [first] ->
+    [
+      WordTildePrefix (strip_tilde first);
+    ]
+  | first :: rest ->
+    [
+      WordTildePrefix (strip_tilde first);
+      WordLiteral ("/" ^ String.concat "/" rest);
+    ]
 
-(** Analyse a [word_component] and lifts the [WordLiteral] that start with a
-    ['~'] in the right places to a [WordTildePrefix] followed by a
-    [WordLiteral]. *)
-let rec make_tilde_prefix_explicit rhs_assignment = function
-  | (WordLiteral s) as cst when s <> "" ->
-    if s.[0] = '~' then (
-      if rhs_assignment then
-        let s = String.split_on_char ':' s in
-        List.(flatten (map find_login s))
-      else
-        find_login s
-    ) else [cst]
-  | WordAssignmentWord (name, Word (s, csts)) ->
-    let csts = recognize ~rhs_assignment:true csts in
-    [WordAssignmentWord (name, Word (s, csts))]
-  | cst ->
-    [cst]
+(** Merges several leading [WordLiteral] into one. *)
+let rec merge_leading_literals : word_cst -> word_cst = function
```

Comment:
Also NIT: I would have accumulated a list of literals first, then use `String.concat` to get the string. Otherwise, you are quadratic in the length of the final string...

Reply:
Rewritten using `Buffer`. cf 02af6a6.